nifi-nar-bundles/nifi-datadog-bundle/nifi-datadog-reporting-task/src/main/java/org/apache/nifi/reporting/datadog/DataDogReportingTask.java

```diff
@@ -0,0 +1,223 @@
+import org.apache.nifi.annotation.documentation.Tags;
+import org.apache.nifi.annotation.lifecycle.OnScheduled;
+import org.apache.nifi.components.AllowableValue;
+import org.apache.nifi.components.PropertyDescriptor;
+import org.apache.nifi.controller.ConfigurationContext;
+import org.apache.nifi.controller.status.ProcessGroupStatus;
+import org.apache.nifi.controller.status.ProcessorStatus;
+import org.apache.nifi.processor.util.StandardValidators;
+import org.apache.nifi.reporting.AbstractReportingTask;
+import org.apache.nifi.reporting.ReportingContext;
+import org.apache.nifi.reporting.datadog.metrics.MetricsService;
+import org.coursera.metrics.datadog.DynamicTagsCallback;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+@Tags({"reporting", "datadog", "metrics"})
+@CapabilityDescription("Publishes metrics from NiFi to datadog")
+public class DataDogReportingTask extends AbstractReportingTask {
+
+    static final AllowableValue DATADOG_AGENT = new AllowableValue("DataDog Agent", "DataDog UDP Agent",
+            "Metrics will be sent via locally installed DataDog agent over UDP");
+
+    static final AllowableValue DATADOG_HTTP = new AllowableValue("Datadog HTTP", "Datadog HTTP",
+            "Metrics will be sent via HTTP transport with no need of Agent installed. " +
+                    "DataDog API key needs to be set");
+
+    static final PropertyDescriptor DATADOG_TRANSPORT = new PropertyDescriptor.Builder()
+            .name("DataDog transport")
+            .description("Transport through which metrics will be sent to DataDog")
+            .required(true)
+            .allowableValues(DATADOG_AGENT, DATADOG_HTTP)
+            .defaultValue(DATADOG_AGENT.getValue())
+            .build();
+
+    static final PropertyDescriptor API_KEY = new PropertyDescriptor.Builder()
+            .name("API key")
+            .description("DataDog API key. If specified value is 'agent', local DataDog agent will be used.")
+            .expressionLanguageSupported(false)
+            .required(false)
+            .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
+            .build();
+
+    static final PropertyDescriptor METRICS_PREFIX = new PropertyDescriptor.Builder()
+            .name("Metrics prefix")
+            .description("Prefix to be added before every metric")
+            .required(true)
+            .expressionLanguageSupported(false)
+            .defaultValue("nifi")
+            .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
+            .build();
+
+    static final PropertyDescriptor ENVIRONMENT = new PropertyDescriptor.Builder()
+            .name("Environment")
+            .description("Environment, dataflow is running in. " +
+                    "This property will be included as metrics tag.")
+            .required(true)
+            .expressionLanguageSupported(false)
```

Comment:
Kinda contrary to a previous comment, I'd suggest making this support EL (and then adding back in the evaluation). This is because with EL it allows the user to access environment variables which may be of use. Sorry about that.

This should probably also be done to METRICS_PREFIX.